### DISCUSSION TITLE
能否自定义一个“虚拟”的package

### DISCUSSION THREAD

**Discussion by Zcaic** [xmake-io/xmake#5714] [Ideas]
### Xmake Version

2.9.5

### Operating System Version and Architecture

windows10

### Describe Bug

xmake能否提供一个虚拟的自定义package。
举例说明一下，比如`package("pybind11")`的代码里面有`add_deps(""python)`。但是此时我想用自己本地的python，而我的路径可能比较复杂，xmake并不能自动找到，那么我就想，我自己知道我的python的头文件路径和库目录，能否自己创建一个“虚拟”的`package("my_python")`，该package在函数`on_install`中并不会真的去安装python，而只是告诉该包的头文件目录和库目录。手动创建一个这样“虚拟”的包后，我在安装`pybind11`时，只需要添加`add_deps("my_python")`，那么pybind就能够找到我本地的python目录。
这只是举一个例子，就是想请问下，是否支持这种只声明相关目录的“虚拟”包，类似于meson的`declare_dependency`，
```
#-------------------meson declare_dependency-------------
my_inc = include_directories(...)
my_lib = static_library(...)
my_dep = declare_dependency(link_with : my_lib,
  include_directories : my_inc)
```
这样的话，包的定义方式就能够有更大的自由度了。

### Expected Behavior

能够手动声明一个本地的package。

### Project Configuration

无

### Additional Information and Error Logs

无

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
用 on_fetch 做个 fetchonly package 就行了。https://github.com/xmake-io/xmake-repo/blob/bbe52197066a7a538fbac43d42139f813d9d5921/packages/o/opengl/xmake.lua#L6

另外，所有的 package 都会自动优先找系统库，走 set_base + on_fetch 重写下查找规则就行了，有没有 on_install 都无所谓。。找到了，就不会走 on_install 逻辑。。也可以 `add_requires("xxx", {system = true})` 强制走系统 不走 on_install

**Reply by Zcaic** [reply]
我改了一下`modules\package\tools\cmake.lua`中的`buildenvs`函数，因为在我使用fetchonly package的时候，发现pybind的cmake不仅需要python的头文件目录和lib目录，还需要python.exe的目录。
![微信截图_20241015002930](https://github.com/user-attachments/assets/c3ce5b42-055c-44a3-ba06-0c7495d05d16)
而我发现在`buildenvs`函数中，查找系统库时，并不会将可执行文件的路径添加到`CMAKE_PREFIX_PATH`。因此，我添加了一句红框里面的命令。
![image](https://github.com/user-attachments/assets/ac8e85ae-77da-4450-8f38-9e0c3146708d)
并在on_fetch的时候，在返回的列表中，增加path值，如下：
![image](https://github.com/user-attachments/assets/2f829249-7b53-4499-a798-26664722d72d)
这样修改以后，pybind能够成功安装，代码片段如下：
![image](https://github.com/user-attachments/assets/4aae7b54-8e34-4d6e-b446-4dd9590bb989)

**Reply by Zcaic** [reply]
@waruqi 这个特性是不是可以提交到源码中，改一下

**Reply by waruqi** [reply]
> 而我发现在buildenvs函数中，查找系统库时，并不会将可执行文件的路径添加到CMAKE_PREFIX_PATH。因此，我添加了一句红框里面的命令。

故意没加，这个干扰太大，这里原本就只是为了让 cmake 找库，不是找 bin 。。

bin 如果在系统 path 里，cmake 原本就能够找到，找不到只能是包的问题，自己打 patch

**Reply by waruqi** [reply]
另外，fetchinfo.path 目前没开放这个 key ，不支持，不要随意乱设置 key 。。自己定义的包用可以，官方仓库和源码 目前不会加这个

**Reply by Zcaic** [reply]
ok好的。

**Comment by Zcaic** [comment]
牛啊，我去试试看，感谢大佬

[a comment by an automated account is not shown]

**Comment by Zcaic** [comment]
还有一个问题，就是我在`package`定义中，使用`add_deps("mypython3")`添加依赖包，发现它默认不会走系统包那一条，即使我的`mypython3` 只是一个fetchonly package，需要我显式的声明`add_deps("mypython3",{system=true})`，才会走系统包那一条路，也就是`package:is_system()`为真。

**Reply by waruqi** [reply]
包内部 deps 走 system 不确定因素太多 ，很容易导致包安装失败，我强制不走系统了。除非你自己 system=false显式指定
